### ISSUE TITLE
Modify on_demand behaviour to fall back to internally generated certificate

### ISSUE THREAD

**Issue by whizzygeeks** [caddyserver/caddy#5627]
If automatic certificate issuance fails in https:// block , we can see a ugly browser error in pic

![image](https://github.com/caddyserver/caddy/assets/32095065/bcb96ee3-43e6-4de0-8677-6550aa541c4e)

Either there should be an option to force it to fallback to fallback certificate/key like in nginx or it should be redirected to port 80. Error handling for on_demand directive is non existent

**Comment by francislavoie** [comment]
If Caddy doesn't have a valid certificate, the TLS handshake can't complete. And browsers won't trust anything from the server without a valid certificate, so redirecting to port 80 is impossible.

**Comment by whizzygeeks** [comment]
I want it to load internal self signed certificate rather then throwing ugly error.  4735 doesnt work for me . Tried mentioning multiple issuers. I am seeing below error

automation policy from site block is also default/catch-all policy because of key without hostname, and the two are in conflict: []certmagic.Issu>r{(*caddytls.ACMEIssuer)(0xc0001168c0), (*caddytls.ZeroSSLIssuer)(0xc00032a7b0), (*caddytls.InternalIssuer)(0xc00032a840)} != []certmagic.Issuer{(*caddytls>

My config looks like below :

https://  {
tls
{
on_demand
}
}

**Comment by mholt** [comment]
Issuer fallback during on-demand doesn't always make sense because on-demand TLS requires low latency to be effective. Sometimes even the first error can take a long time to reach.

> automation policy from site block is also default/catch-all policy because of key without hostname, and the two are in conflict: []certmagic.Issu>r{(*caddytls.ACMEIssuer)(0xc0001168c0), (*caddytls.ZeroSSLIssuer)(0xc00032a7b0), (*caddytls.InternalIssuer)(0xc00032a840)} != []certmagic.Issuer{(*caddytls>

This error is truncated, and I'm highly skeptical it actually came from the config you posted. We need the _full_ unredacted config as well as the full error messages to be most productive. Here's a template to help us get the info we need: 

It's not immediately clear to me what is going on, so I'll need your help to understand it better.

Ideally, we need to be able to reproduce the bug _in the most minimal way possible_ **using the latest version of Caddy.** This allows us to write regression tests to verify the fix is working. If we can't reproduce it, then you'll have to test our changes for us until it's fixed -- and then we can't add test cases, either.

I've attached a template below that will help make this easier and faster! **This will require some effort on your part -- please understand that we will be dedicating time to fix the bug you are reporting if you can just help us understand it and reproduce it easily.**

This template will ask for some information you've already provided; that's OK, just fill it out the best you can. :+1: I've also included some helpful tips below the template. Feel free to let me know if you have any questions!

Thank you again for your report, we look forward to resolving it!

## Template

<pre><code>## 1. Environment

### 1a. Operating system and version

```
paste here
```


### 1b. Caddy version (run `caddy version` or paste commit SHA)

This should be the latest version of Caddy:

```
paste here
```


## 2. Description

### 2a. What happens (briefly explain what is wrong)




### 2b. Why it's a bug (if it's not obvious)




### 2c. Log output

```
paste terminal output or logs here
```



### 2d. Workaround(s)




### 2e. Relevant links




## 3. Tutorial (minimal steps to reproduce the bug)




</code></pre>

## Instructions -- please heed otherwise we cannot help you (help us help you!)

1. **Environment**: Please fill out your OS and Caddy versions, even if you don't think they are relevant. (They are _always_ relevant.) If you built Caddy from source, provide the commit SHA and specify your exact Go version.

2. **Description:** Describe at a high level what the bug is. What happens? Why is it a bug? Not all bugs are obvious, so convince readers that it's actually a bug.
	- **2c) Log output:** Paste terminal output and/or complete logs in a code block. [DO NOT REDACT INFORMATION except for credentials.](https://caddy.community/t/how-to-get-help-with-caddy-more-effectively/5222) Please enable debug and access logs.
	- **2d) Workaround:** What are you doing to work around the problem in the meantime? This can help others who encounter the same problem, until we implement a fix.
	- **2e) Relevant links:** Please link to any related issues, pull requests, docs, and/or discussion. This can add crucial context to your report.

3. **Tutorial:** What are the _minimum required specific steps_ someone needs to take in order to experience the same bug? Your goal here is to make sure that anyone else can have the same experience with the bug as you do. You are writing a tutorial, so make sure to carry it out yourself before posting it. Please:
	- Start with an empty config. Add _only_ the lines/parameters that are _absolutely required_ to reproduce the bug.
	- Do not run Caddy inside containers.
	- Run Caddy manually in your terminal; do not use systemd or other init systems.
	- If making HTTP requests, avoid web browsers. Use a simpler HTTP client instead, like `curl`.
	- Do not redact any information from your config (except credentials). Domain names are public knowledge and [often necessary for quick resolution of an issue](https://caddy.community/t/how-to-get-help-with-caddy-more-effectively/5222)!
	- Note that ignoring this advice may result in delays, or even in your issue being closed. 😞 Only actionable issues are kept open, and if there is not enough information or clarity to reproduce the bug, then the report is not actionable.

Example of a tutorial:

<blockquote>Create a config file:

```
{ ... }
```

Open terminal and run Caddy:

```
$ caddy ...
```

Make an HTTP request:

```
$ curl ...
```

Notice that the result is ___ but it should be ___.</blockquote>

**Comment by mholt** [comment]
Ok, yeah so thanks for sharing the full config, now it's obvious. The error is saying that the "automation policy from site block is also default/catch-all policy because of key without hostname" -- referring to your site block, `https://` -- there is no hostname, and it has a `tls` directive which is configuring the automation policy (issuers in this case). This is in conflict with the settings for TLS automation that you specified in your global options. They have to match exactly (which doesn't make sense to have both anyway, usually) because they're both saying "This is how you manage certificates for _everything other than the configured hostnames with their own automation policies."

Basically, remove either the global options related to TLS automation, or remove the `tls` block from your `https://` site.

**Comment by francislavoie** [comment]
Also, your config doesn't really make sense, there's _a lot_ of redundancies. Your entire site block can be replaced with just this:

```
https:// {
	tls {
		on_demand
	}

	reverse_proxy internal-PROD-example-THEME-INTERNAL-ALB-000000000.us-east-1.elb.amazonaws.com:80
}
```

**Comment by whizzygeeks** [comment]
If we use on_demand it will show ugly ssl error page in case of tls issuance failure. This is the reason why i have raised this issue. There is no option of using fallback internal certificates or showing custom cert error or redirecting to port 80. 
Users do not like unknown error on site in case due to some reason tls generation fails , they rather will prefer content with tls warning or proper error message.

Note: Multiple directives for reverse proxy is due to the fact we have separate backends based on post and get http methods

**Comment by mholt** [comment]
I understand; however:

> There is no option of using fallback internal certificates

This will still show an ugly SSL error to your users, unless they have installed your root cert first.

> or showing custom cert error

I don't think custom error messages are a thing with TLS. (I could be wrong but I'd be surprised if this was a practical feature.)

> or redirecting to port 80.

A redirect is impossible because that's an HTTP feature, but we can't send HTTP until a connection is established.

Browsers often try port 80 automatically if a connection to port 443 fails.

> Note: Multiple directives for reverse proxy is due to the fact we have separate backends based on post and get http methods

But this line is repeated 3 times:

```
reverse_proxy @post internal-PROD-example-THEME-INTERNAL-ALB-000000000.us-east-1.elb.amazonaws.com:80 {
```

And most, if not all, of the `header_up` lines are unnecessary.

**Comment by francislavoie** [comment]
I'll close this - I don't see anything actionable here, except for possibly serving `internal` certs as a fallback, but there's already https://github.com/caddyserver/caddy/issues/4735 open for that, as I mentioned earlier.
